### DISCUSSION TITLE
Add support for server-timing header

### DISCUSSION THREAD

**Discussion by timneutkens** [vercel/next.js#62353] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

The Server-Timing header allows for showing time traces in the browser devtools for what happened when a page was rendering, this can be useful to show the time it took to run getStaticProps/getStaticPaths for example.

Some references:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Server-Timing
https://twitter.com/paul_irish/status/829090506084749312

## Describe the solution you'd like

Ideally we'd enable this automatically in development. In production it should probably be under a flag.

## Additional context

Reference issue: #7035

**Comment by cwohlman** [comment]
Has anyone taken this issue yet?

**Comment by asci** [comment]
that might be quite handy for debugging long TTFB

**Comment by ghost** [comment]
>  this can be useful to show the time it took to run getStaticProps/getStaticPaths for example

Do you mean `getServerSideProps` ? Maybe I understand it wrong but this is just an extra header in the response. Not sure how this could work for `getStaticProps`?

**Comment by timneutkens** [comment]
@roelandmoors during development each request triggers getStaticPaths/getStaticProps to be run so it can be measured there and surfaced.

**Comment by jasonwilliams** [comment]
When I was using React with express, one of the things I was able to do was wrap `renderToString` in a performance timer then measure how long react was taking to render pages. With Next I’m unable to do this now as that isn’t exposed.

i already use the server-timing headers in my app so as long as next doesn’t conflict with that it should be ok. I wonder if instead they could use the performance API and expose the results some how.

**Comment by mohsen1** [comment]
So I have implemented this in user space for our Next.js deployment internally.  This is totally useful to measure what's slow in prop generating server side functions like `getServerSideProps`. 

What's sad about this is that the headers do not have time stamps and they are just durations. So you don't get a nice waterfall or flamechart of what has happened. I feel this standard needs a few additional features to be perfectly informative. The headers need to also have offset time so we can see the waterfall. There should be a way of telling which span is within another span to build flamecharts.


Example server response headers in our case
```
server-timing: query.GetViewer;dur=198
server-timing: query.Page_home;dur=88
server-timing: React.render;dur=114
server-timing: Document.getInitialProps;dur=68
```
<img width="646" alt="Screen Shot 2022-08-20 at 1 29 10 PM" src="https://user-images.githubusercontent.com/543633/185759470-c85ff514-c4c0-4c17-82f0-210f157ef64e.png">


P.S. [Almost got scammed getting to here](https://twitter.com/mohsen____/status/1561043767616045056)!

**Comment by jasonwilliams** [comment]
@mohsen1 I agree with you it would be nice to have server timing show a waterfall chart (similar to what is seen above in request/response). I guess it would take reaching out to whoever worked on Server Timings API (@igrigorik ?) to see if its possible to add on timestamps to each timing.

I don't know if there's future work to have Server Timing use [Entry Types](https://w3c.github.io/perf-timing-primer/#entrytype) which do use start-time and duration. It seems the effort may have been abandoned but that would have helped map [PerformanceEntry](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceEntry) directly to a ServerTiming entry.

**Comment by steve-marmalade** [comment]
I came here looking for this exact feature. As we move more and more of the business logic to the server, having improved server-side telemetry will increasingly help us understand our applications and debug performance issues.

**Comment by igorgatis** [comment]
It would be even cooler if could easily connect it with something like https://github.com/wallet77/v8-inspector-api#cpu-profiling

**Comment by galaxynova1999** [comment]
really need this feature

**Comment by jodosha** [comment]
Any news?

**Comment by FranCarstens** [comment]
We also have this need. Any word from Vercel on this would be appreciated.

**Comment by Svish** [comment]
Would be really awesome if this was just a config I could enable in my `next.config.js`, and get useful `Server-Timing` of whatever would make sense from Next.js rendering my pages.

Should probably be always on by default in development, and opt-in for production via config in `next.config.js`.

**Comment by viktorbezdek** [comment]
We'd love to have this feature too.

**Comment by xiangxu-meng** [comment]
Is there currently a method to obtain the time consumption of server-side renderToString?

**Comment by xiangxu-meng** [comment]
> 因此，我在内部 Next.js 部署的用户空间中实现了这一点。这对于测量 prop 生成服务器端函数（例如`getServerSideProps`.
> 
> 令人遗憾的是，标头没有时间戳，它们只是持续时间。所以你不会得到一个关于所发生的事情的漂亮的瀑布图或火焰图。我觉得这个标准需要一些额外的功能才能提供完整的信息。标头还需要有偏移时间，以便我们可以看到瀑布。应该有一种方法来判断哪个跨度位于另一个跨度内以构建火焰图。
> 
> 在我们的例子中，服务器响应标头示例
> 
> ```
> server-timing: query.GetViewer;dur=198
> server-timing: query.Page_home;dur=88
> server-timing: React.render;dur=114
> server-timing: Document.getInitialProps;dur=68
> ```
> 
> <img alt="屏幕截图 2022 年 8 月 20 日下午 1 点 29 分" width="646" src="https://user-images.githubusercontent.com/543633/185759470-c85ff514-c4c0-4c17-82f0-210f157ef64e.png">
> 
> PS[来到这里差点被骗了](https://twitter.com/mohsen____/status/1561043767616045056)！

How do you get the time of React.render?

**Comment by lforst** [comment]
This would be incredibly useful for monitoring services to be able to associate SSR requests with frontend timings.

**Comment by sadeghmohebbi** [comment]
also web.dev google mentioned in this article: https://web.dev/articles/optimize-ttfb#understanding_high_ttfb_with_server-timing

**Comment by samcx** [comment]
Hi everyone!

Since this is a feature request, I will be moving this over to <picture data-single-emoji=":nextjs:" title=":nextjs:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/nextjs/2a85d633ae594556.png" alt=":nextjs:" align="absmiddle"></picture> discussions.

**Reply by SilentImp** [reply]
Any updates on this feature?

**Reply by meetmatt** [reply]
++

**Reply by lforst** [reply]
It's not exactly the same as `server-timing` support, but I opened a PR that should pretty much cover all the use-cases since you could propagate arbitrary metadata to the client (similar to a header) which is readable through JS: https://github.com/vercel/next.js/pull/64256

**Comment by HananoshikaYomaru** [comment]
so `Server-Timing` response headers is completely removed from vercel deployment? I tried deploy a boilerplate to vercel but server timing is not working.

**Comment by tony13tv** [comment]
I have a prototype using AsyncLocalStorage. It allows calling `metrics()` to obtain a map inside server actions to insert metrics with the correspondent duration and description. Automatically being assigned to the response header 'Server-Timing' during the `action` phase of the requestStore.

```typescript
"use server";
import { metrics } from "next/server";
import { db } from "./drizzle"

export async function getUsers(search?: string) {
  const m = metrics(); // <- type here would be Map<string, { dur?: string, desc?: string }> adhering to reference. https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Server-Timing#directives
  const time = performance.now();
  const data = await db.query.user.findMany();
  m.set("db(getUsers)", { dur: performance.now() - time, desc: "time spent retrieving all the users"})
  return data;
}
```

If anyone thinks this is useful let me know to open a PR.

**Reply by tony13tv** [reply]
You can check the PR out [here](https://github.com/vercel/next.js/pull/78608)

**Comment by tony13tv** [comment]
For anyone looking for a solution can checkout this [library](https://github.com/prograpedia/next-server-timing). It has been tested with the latest version of next.

**Comment by shogunsea** [comment]
it's so ridiculous 5 years later the most popular frontend hosting provider still has no support for this standard performance debug header, not even disclosing the reason why vercel didn't plan to add support for this.

**Comment by rgembalik** [comment]
This is very much needed, especially for dynamic pages. I would love to see timing split between response generation (html or json for rsc), fetches and possibly internal modules.
